File: apps/macos/Sources/MyJobFinder/WorkerClient.swift
import Foundation

final class WorkerClient {
    private let workspace: URL
    private let decoder = JSONDecoder()
    private let encoder = JSONEncoder()

    init(workspace: URL? = nil) throws {
        self.workspace = try workspace ?? WorkspaceLocator.find()
    }

    func state() async throws -> WorkerEnvelope {
        try await run(command: "state")
    }

    func configure(_ input: DesktopProfileInput) async throws -> WorkerEnvelope {
        try await run(command: "configure", input: try encoder.encode(input))
    }

    func refresh(onEvent: @escaping @Sendable (WorkerEvent) -> Void = { _ in }) async throws -> WorkerEnvelope {
        try await run(command: "refresh", onEvent: onEvent)
    }

    private func run(command: String, input: Data? = nil, onEvent: @escaping @Sendable (WorkerEvent) -> Void = { _ in }) async throws -> WorkerEnvelope {
        try await Task.detached {
            let worker = self.workspace.appendingPathComponent("dist/desktop/cli.js")
            guard FileManager.default.fileExists(atPath: worker.path) else {
                throw CocoaError(.fileNoSuchFile, userInfo: [NSLocalizedDescriptionKey: "Run npm run desktop:build-worker before opening My Job Finder."])
            }

            let process = Process()
            let bundledNode = self.workspace.appendingPathComponent("node")
            if FileManager.default.fileExists(atPath: bundledNode.path) {
                process.executableURL = bundledNode
                process.arguments = [worker.path, command]
            } else {
                process.executableURL = URL(fileURLWithPath: "/usr/bin/env")
                process.arguments = ["node", worker.path, command]
            }
            process.currentDirectoryURL = self.workspace
            var environment = ProcessInfo.processInfo.environment
            let appData = FileManager.default.homeDirectoryForCurrentUser.appendingPathComponent("Library/Application Support/My Job Finder").path
            environment["MY_JOB_FINDER_HOME"] = appData
            environment["MY_JOB_FINDER_ENGINE_HOME"] = appData
            let bundledBrowsers = self.workspace.appendingPathComponent("ms-playwright").path
            if FileManager.default.fileExists(atPath: bundledBrowsers) {
                environment["PLAYWRIGHT_BROWSERS_PATH"] = bundledBrowsers
            }
            process.environment = environment
            let stdout = Pipe()
            let stderr = Pipe()
            let stdin = Pipe()
            process.standardOutput = stdout
            process.standardError = stderr
            process.standardInput = stdin
            try process.run()
            if let input { stdin.fileHandleForWriting.write(input) }
            try? stdin.fileHandleForWriting.close()
            let outputData = stdout.fileHandleForReading.readDataToEndOfFile()
            let errorData = stderr.fileHandleForReading.readDataToEndOfFile()
            process.waitUntilExit()

            var final: WorkerEnvelope?
            let lines = String(decoding: outputData, as: UTF8.self).split(whereSeparator: \.isNewline)
            for line in lines {
                let envelope = try self.decoder.decode(WorkerEnvelope.self, from: Data(line.utf8))
                if let event = envelope.event { onEvent(event) }
                if envelope.state != nil || envelope.error != nil || envelope.profile != nil { final = envelope }
            }
            guard process.terminationStatus == 0, let final, final.ok != false else {
                let stderrText = String(decoding: errorData, as: UTF8.self)
                throw CocoaError(.executableRuntimeMismatch, userInfo: [NSLocalizedDescriptionKey: final?.error ?? stderrText])
            }
            return final
        }.value
    }
}

File: apps/windows/MyJobFinder/Services/WorkerClient.cs
using System.Diagnostics;
using MyJobFinder.Core;

namespace MyJobFinder.Services;

public sealed class WorkerClient
{
    private readonly string workspaceRoot;
    private readonly string workerPath;
    private readonly string nodePath;
    private readonly string appDataRoot;

    public WorkerClient()
    {
        workspaceRoot = WorkspaceLocator.Find(AppContext.BaseDirectory);
        workerPath = Path.Combine(workspaceRoot, "dist", "desktop", "cli.js");
        nodePath = File.Exists(Path.Combine(workspaceRoot, "node.exe"))
            ? Path.Combine(workspaceRoot, "node.exe")
            : "node.exe";
        appDataRoot = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData),
            "My Job Finder");
        if (!File.Exists(workerPath))
        {
            throw new FileNotFoundException("Build the shared worker with npm run desktop:build-worker.", workerPath);
        }
    }

    public Task<WorkerEnvelope> GetStateAsync() => RunAsync("state", null, null);

    public Task<WorkerEnvelope> ConfigureAsync(DesktopProfileInput input)
        => RunAsync("configure", WorkerProtocol.SerializeProfile(input), null);

    public Task<WorkerEnvelope> RefreshAsync(IProgress<WorkerEvent>? progress = null)
        => RunAsync("refresh", null, progress);

    private async Task<WorkerEnvelope> RunAsync(string command, string? input, IProgress<WorkerEvent>? progress)
    {
        var startInfo = new ProcessStartInfo
        {
            FileName = nodePath,
            WorkingDirectory = workspaceRoot,
            RedirectStandardInput = true,
            RedirectStandardOutput = true,
            RedirectStandardError = true,
            UseShellExecute = false,
            CreateNoWindow = true
        };
        startInfo.ArgumentList.Add(workerPath);
        startInfo.ArgumentList.Add(command);
        startInfo.Environment["MY_JOB_FINDER_HOME"] = appDataRoot;
        startInfo.Environment["MY_JOB_FINDER_ENGINE_HOME"] = appDataRoot;
        var bundledBrowsers = Path.Combine(workspaceRoot, "ms-playwright");
        if (Directory.Exists(bundledBrowsers))
        {
            startInfo.Environment["PLAYWRIGHT_BROWSERS_PATH"] = bundledBrowsers;
        }

        using var process = new Process { StartInfo = startInfo };
        process.Start();
        if (input is not null) await process.StandardInput.WriteAsync(input);
        process.StandardInput.Close();
        var errorTask = process.StandardError.ReadToEndAsync();
        WorkerEnvelope? final = null;

        while (await process.StandardOutput.ReadLineAsync() is { } line)
        {
            if (string.IsNullOrWhiteSpace(line)) continue;
            var envelope = WorkerProtocol.ParseEnvelope(line);
            if (envelope.Event is not null) progress?.Report(envelope.Event);
            if (envelope.State is not null || !envelope.Ok || envelope.Profile is not null) final = envelope;
        }

        await process.WaitForExitAsync();
        var error = await errorTask;
        if (process.ExitCode != 0 || final is null || !final.Ok)
        {
            throw new InvalidOperationException(final?.Error ?? error.Trim() ?? "The job worker failed.");
        }
        return final;
    }
}
